#159053980 Feature JWT Authentication

davenjeru opened the pull request:
**What does this PR do?**
Implements a simple version of authenticating users using JSON Web Tokens.

**Description of Task to be completed?**
- Implement a method for generating a JSON Web Token when a user registers or logs in.
- Implement a method for authenticating tokens received from requests to protected endpoints.

**How should this be manually tested?**
- Clone the repository from [here](https://github.com/andela/ah-scorpion/tree/ft-JWT-Auth-%23159053980) and check out the branch `ft-JWT-Auth-#159053980`
- Activate your virtual environment and use the `.env-sample` to set the environment variables.
- Use postman to test these endpoints.
  `POST api/users` for registering
  `POST api/users/login` for logging in
When a user successfully registers or logs in, a JWT should be sent in the response.

- To test whether the token is valid, copy and paste it into the `Authorization` header and prepend it with the keyword `'Bearer'` then access this endpoint:
  `GET api/user` If successful, the response should be the user's details, i.e, the username and email.

What are the relevant pivotal tracker stories?
[#159053980](https://www.pivotaltracker.com/story/show/159053980)

davenjeru committed:
[feature #159053980] JWT Authentication
 - make migrations
 - add app_name to authentication/urls.py

davenjeru committed:
[feature #159053980] JWT Authentication
 - implement method for generating tokens
 - add JWTAuthentication class

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[feature #159053980] JWT Authentication
 - remove print debug statements

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[feature #159053980] JWT Authentication
 - implement method for authenticating JWT received from request

Signed-off-by: davenjeru <dave.njeru@andela.com>

RuthNjeri commented on authors/apps/authentication/views.py:
```diff
@@ -44,7 +51,20 @@ def post(self, request):
         serializer.is_valid(raise_exception=True)
 
-        return Response(serializer.data, status=status.HTTP_200_OK)
+        token = generate_token(serializer.data).decode()
```
I noticed that the token generated for login and registration are the same. If you include a payload with details such as the time the token is generated, it will generate different tokens for each. You can also include expiry inside it.
` payload = {
            'exp': datetime.datetime.utcnow() + datetime.timedelta(days=0, seconds=5),
            'iat': datetime.datetime.utcnow(),
            'sub': output
        }`

dmigwi replied:
Great feedback. 👏 @RuthNjeri

davenjeru replied:
Okay, I will put the issuing time and expiry time in the payload to be encoded by 9.30 a.m.
Thank you @RuthNjeri

dmigwi commented on authors/apps/authentication/backends.py:
```diff
@@ -1,10 +1,24 @@
+
+from .models import User
 
 """Configure JWT Here"""
```
Delete this comment since you've configured JWT

davenjeru replied:
Okay...

dmigwi reviewed (commented):
Progress looks wonderful

hoslack commented on authors/apps/authentication/views.py:
```diff
@@ -44,7 +51,20 @@ def post(self, request):
         serializer.is_valid(raise_exception=True)
 
-        return Response(serializer.data, status=status.HTTP_200_OK)
+        token = generate_token(serializer.data).decode()
```
@davenjeru That's a great PR, I just noticed that you haven't included the instructions on how to modify the .env file to add the secret key. This would help in testing out your work.

davenjeru replied:
The secret key is configured in [authors/settings.py](https://github.com/andela/ah-scorpion/blob/develop/authors/settings.py#L22).
The app should work without changing your `.env` file.
For security reasons, I'll set the Secret Key to be read from an environment variable before 9.30 am today.

RuthNjeri commented on authors/apps/authentication/views.py:
```diff
@@ -1,9 +1,11 @@
 from rest_framework.response import Response
 from rest_framework.views import APIView
 
+from authors.settings import SECRET_KEY
```
@hoslack the secret key has been imported here from the settings file. I do not think @davenjeru needs to create a .env file for that.

davenjeru committed:
[feature #159053980] JWT Authentication
 - SECRET_KEY now obtained from environment variable 'SECRET_KEY'

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[feature #159053980] JWT Authentication
 - add expiry time to the generated token. Expiry time set to 1 day

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[feature #159053980] JWT Authentication
 - add keyword = 'Bearer' to be used in the header instead of 'Token'
 - remove the 'authenticate' function. We do not need to override this
   function when keyword is set to 'Bearer'
 - override the 'authenticate_credentials' function to check for token
   validity then return user object.

Signed-off-by: davenjeru <dave.njeru@andela.com>

munyili-samuel commented:
I have tested the code and it is working well.
Good job. 👍

davenjeru commented on .env-sample:
```diff
@@ -3,3 +3,4 @@ export USER="db_user"
 export PASSWORD="db_password"
 export HOST="db_host"
 export PORT="db_port"
+export SECRET_KEY = 'YOUR SECRET KEY'
```
@hoslack here is the SECRET KEY setting on the .env sample

RuthNjeri reviewed (approved):
Good work on the improvements. I have tested it and everything works fine.

davenjeru committed:
[feature #159053980] JWT Authentication
 - add 'errors' key on response where this was not displayed properly

Signed-off-by: davenjeru <dave.njeru@andela.com>

lennykioko commented on authors/settings.py:
```diff
@@ -19,7 +19,7 @@
 
 # SECURITY WARNING: keep the secret key used in production secret!
-SECRET_KEY = '7pgozr2jn7zs_o%i8id6=rddie!*0f0qy3$oy$(8231i^4*@u3'
+SECRET_KEY = os.environ.get('SECRET_KEY')
```
Use `os.environ["SECRET_KEY"]` instead. This will raise an error if the secret key is not provided from the environment variable. When you use `os.environ.get("SECRET_KEY")` it returns **None** when the key is not present

davenjeru replied:
Okay. Let me implement these changes.

zakhusayn commented on authors/apps/authentication/views.py:
```diff
@@ -44,7 +53,24 @@ def post(self, request):
+    Method that generates a JSON Web Token for the user
+    :param identity: User information to be encoded as a dictionary
+    :return: JWT token
+    :rtype: bytes
```
spelling error: rtype instead of type

davenjeru replied:
:rtype: stands for 'return type' which in this case is bytes.

zakhusayn commented on authors/apps/authentication/views.py:
```diff
@@ -44,7 +53,24 @@ def post(self, request):
+    :return: JWT token
+    :rtype: bytes
+    """
+    payload = dict(identity=identity,
```
Instead of passing all user's information into the payload. Try passing email or user_id instead

davenjeru replied:
We pass in the email and username only. That is what is in the identity.

davenjeru committed:
[feature #159053980] JWT Authentication
 - remove space from the '=' sign.

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[feature #159053980] JWT Authentication
 - get SECRET_KEY from environment by subscripting so that it raises
   an error if it is not set.

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru commented on authors/settings.py:
```diff
@@ -19,7 +19,7 @@
 
 # SECURITY WARNING: keep the secret key used in production secret!
-SECRET_KEY = '7pgozr2jn7zs_o%i8id6=rddie!*0f0qy3$oy$(8231i^4*@u3'
+SECRET_KEY = os.environ['SECRET_KEY']
```
@lennykioko Here is the change
